gui/textbox.3
.\"	$Csoft: textbox.3,v 1.28 2005/04/25 02:19:52 vedge Exp $
.\"
.\" Copyright (c) 2002, 2003, 2004, 2005 CubeSoft Communications, Inc.
.\" <http://www.csoft.org>
.\" All rights reserved.
.\"
.\" Redistribution and use in source and binary forms, with or without
.\" modification, are permitted provided that the following conditions
.\" are met:
.\" 1. Redistributions of source code must retain the above copyright
.\"    notice, this list of conditions and the following disclaimer.
.\" 2. Redistributions in binary form must reproduce the above copyright
.\"    notice, this list of conditions and the following disclaimer in the
.\"    documentation and/or other materials provided with the distribution.
.\" 
.\" THIS SOFTWARE IS PROVIDED BY THE AUTHOR ``AS IS'' AND ANY EXPRESS OR
.\" IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED
.\" WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
.\" ARE DISCLAIMED. IN NO EVENT SHALL THE AUTHOR BE LIABLE FOR ANY DIRECT,
.\" INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES
.\" (INCLUDING BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR
.\" SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION)
.\" HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT,
.\" STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING
.\" IN ANY WAY OUT OF THE USE OF THIS SOFTWARE EVEN IF ADVISED OF THE
.\" POSSIBILITY OF SUCH DAMAGE.
.\"
.Dd August 21, 2002
.Dt TEXTBOX 3
.Os
.ds vT Agar API Reference
.ds oS Agar 1.0
.Sh NAME
.Nm textbox
.Nd agar text box widget
.Sh SYNOPSIS
.Bd -literal
#include <agar/gui/textbox.h>
.Ed
.Sh DESCRIPTION
The
.Nm
widget binds to a NUL-terminated string and allows the user to edit its
contents.
The string may contain UTF-8 encoded characters.
.Sh INHERITANCE HIERARCHY
.Pp
.Xr AG_Object 3 ->
.Xr widget 3 ->
.Nm .
.Sh INITIALIZATION
.nr nS 1
.Ft "AG_Textbox *"
.Fn AG_TextboxNew "void *parent" "Uint flags" "const char *label"
.Pp
.Ft void
.Fn AG_TextboxInit "AG_Textbox *textbox" "Uint flags" "const char *label"
.Pp
.Ft void
.Fn AG_TextboxSetWriteable "AG_Textbox *textbox" "int writeable"
.Pp
.Ft void
.Fn AG_TextboxSetPassword "AG_Textbox *textbox" "int password"
.Pp
.Ft void
.Fn AG_TextboxPrescale "AG_Textbox *textbox" "const char *text"
.nr nS 0
.Pp
The
.Fn AG_TextboxNew
function allocates, initializes, and attaches a
.Nm
widget.
The
.Fn AG_TextboxInit
function initializes a
.Nm
widget with the label text set to
.Fa label .
.Pp
The
.Fn AG_TextboxSetWriteable
function controls the writeability of the string.
.Pp
.Fn AG_TextboxSetPassword
controls password (no print) input.
.Pp
.Fn AG_TextboxPrescale
requests that the initial geometry of
.Fa textbox
is to be sufficient to display the string
.Fa text
in its entirety.
.Sh TEXT MANIPULATION
.nr nS 1
.Ft void
.Fn AG_TextboxPrintf "AG_Textbox *textbox" "const char *fmt" "..."
.Pp
.Ft "char *"
.Fn AG_TextboxDupString "AG_Textbox *textbox"
.Pp
.Ft "size_t"
.Fn AG_TextboxCopyString "AG_Textbox *textbox" "char *dst" "size_t dst_size"
.Pp
.Ft int
.Fn AG_TextboxInt "AG_Textbox *textbox"
.nr nS 0
.Pp
The
.Fn AG_TextboxPrintf
function uses
.Xr vsnprintf 3
to overwrite the contents of the text buffer.
If the
.Fa fmt
argument is NULL, a NUL string is assigned instead.
.Pp
The
.Fn AG_TextboxDupString
function returns a copy of the text buffer.
.Pp
The
.Fn AG_TextboxCopyString
function copies up to
.Fa dst_size
- 1 bytes from the text buffer
.Fa dst ,
NUL-terminating the result and returning the number of bytes that would
have been copied if
.Fa dst_size
was unlimited.
.Pp
.Fn AG_TextboxInt
tries to convert the string to an integer value.
.Sh BINDINGS
The
.Nm
widget provides the following bindings:
.Pp
.Bl -tag -compact -width "char *string "
.It Va char *string
Fixed-size buffer of UTF-8 text.
.El
.Sh EVENTS
The
.Nm
widget reacts to the following events:
.Pp
.Bl -tag -compact -width 25n
.It window-mousebuttondown
Move focus to the widget.
Position the cursor at a specific position.
.It window-mousemotion
Move the cursor and scroll.
.It window-keydown
Perform the action that the current keymap associates with this key.
.El
.Pp
The
.Nm
widget generates the following events:
.Pp
.Bl -tag -width 2n
.It Fn textbox-return "void"
Return was pressed.
.It Fn textbox-prechg "void"
The string is about to be modified.
.It Fn textbox-postchg "void"
The string was just modified.
.El
.Sh EXAMPLES
The following code fragment binds a
.Nm
to a string contained in a fixed-size buffer:
.Bd -literal -offset indent
char name[32];
AG_Textbox *tb;

tb = AG_TextboxNew(parent, 0, "Name: ");
AG_WidgetBind(tb, "string", AG_WIDGET_STRING, name, sizeof(name));
.Ed
.Pp
As is the case with all widgets, a default binding (residing in the widget
structure) is provided.
The following code uses
.Fn AG_TextboxPrintf
to set the value of the default binding, and
.Fn AG_TextboxDupString
to retrieve a pointer to it:
.Bd -literal -offset indent
AG_Textbox *tb;
char *value_ptr;

tb = AG_TextboxNew(parent, 0, "Value: ");
AG_TextboxPrintf(tb, "Foo");
value_ptr = AG_TextboxDupString(tb);
.Ed
.Pp
Note that the default binding is limited to
.Dv AG_TEXTBOX_STRING_MAX
bytes in size.
.Sh SEE ALSO
.Xr agar 3 ,
.Xr text 3 ,
.Xr tlist 3 ,
.Xr widget 3 ,
.Xr window 3
.Sh HISTORY
The
.Nm
widget first appeared in Agar 1.0.
.Sh BUGS
.Fn AG_TextboxInt
will be obsoleted by integral type bindings.
.Pp
UTF-8 string boundary check is not space efficient.


demos/textbox/textbox.c
/*	$Csoft: gamemenu.c,v 1.6 2005/10/07 07:09:35 vedge Exp $	*/
/*	Public domain	*/

/*
 * This application demonstrates the use of the AG_Textbox widget for
 * editing a string.
 */

#include <agar/core.h>
#include <agar/gui.h>

#include <string.h>
#include <unistd.h>

static char text_buf[128];

static void
return_pressed(AG_Event *event)
{
	AG_Textbox *textbox = AG_SELF();
	char my_string[128];

	/*
	 * Copy the current textbox string to a sized buffer. We could also
	 * have used AG_TextboxDupString() to get a malloc'd copy.
	 */
	AG_TextboxCopyString(textbox, my_string, sizeof(my_string));
	AG_TextMsg(AG_MSG_INFO, "Entered string: `%s'", my_string);
}

static Uint32
update_text_buffer(void *obj, Uint32 ival, void *arg)
{
	snprintf(text_buf, sizeof(text_buf), "Ticks: %lu",
	    (unsigned long)SDL_GetTicks());
	return (ival);
}

static void
CreateTextbox(void)
{
	AG_Timeout update_timer;
	AG_Window *win;
	AG_Button *btn;
	AG_Textbox *textbox;
	int i;

	win = AG_WindowNew(0);
	AG_WindowSetCaption(win, "Textbox Example");

	/*
	 * Create a textbox with a label displaying "Enter string:". Use the
	 * `textbox-return' event to catch the return key.
	 */
	textbox = AG_TextboxNew(win, AG_TEXTBOX_HFILL|AG_TEXTBOX_FOCUS,
	    "Enter string: ");
	AG_SetEvent(textbox, "textbox-return", return_pressed, NULL);

	/*
	 * Use a `string' binding to edit the contents of a sized buffer. We
	 * have to pass the size of the buffer along with a pointer to it to
	 * the AG_WidgetBind() function.
	 */
	textbox = AG_TextboxNew(win, AG_TEXTBOX_HFILL, "Edit buffer: ");
	AG_WidgetBind(textbox, "string", AG_WIDGET_STRING, text_buf,
	    sizeof(text_buf));

	/* Use a timer to update the text buffer periodically. */
	snprintf(text_buf, sizeof(text_buf), "Text buffer contents");
	AG_SetTimeout(&update_timer, update_text_buffer, NULL, 0);
	AG_AddTimeout(textbox, &update_timer, 250);

	AG_LabelNew(win, AG_LABEL_POLLED, "Buffer contents: `%s'", &text_buf);

	AG_WindowShow(win);
}

static void
UpdateTable(AG_Event *event)
{
	AG_Table *t = AG_SELF();
	static int prev = 0;
	static int dir = +1;
	int i;

	AG_TableBegin(t);
	for (i = 0; i < prev; i++) {
		AG_TableAddRow(t, "%d:%u", i, (unsigned)SDL_GetTicks());
	}
	AG_TableEnd(t);

	if (dir < 0) {
		if (--prev < 0) { prev = 0; dir = +1; }
	} else {
		if (++prev > 100) { prev = 100; dir = -1; }
	}
}

int
main(int argc, char *argv[])
{
	int c, i, fps = -1;
	char *s;

	if (AG_InitCore("table-demo", 0) == -1) {
		fprintf(stderr, "%s\n", AG_GetError());
		return (1);
	}

	while ((c = getopt(argc, argv, "?vfFgGr:")) != -1) {
		extern char *optarg;

		switch (c) {
		case 'v':
			exit(0);
		case 'f':
			AG_SetBool(agConfig, "view.full-screen", 1);
			break;
		case 'F':
			AG_SetBool(agConfig, "view.full-screen", 0);
			break;
#ifdef HAVE_OPENGL
		case 'g':
			AG_SetBool(agConfig, "view.opengl", 1);
			break;
		case 'G':
			AG_SetBool(agConfig, "view.opengl", 0);
			break;
#endif
		case 'r':
			fps = atoi(optarg);
			break;
		case '?':
		default:
			printf("%s [-vfFgG] [-r fps]\n", agProgName);
			exit(0);
		}
	}

	/* Initialize a 640x480x32 display. Respond to keyboard/mouse events. */
	if (AG_InitVideo(640, 480, 32, 0) == -1 ||
	    AG_InitInput(0) == -1) {
		fprintf(stderr, "%s\n", AG_GetError());
		return (-1);
	}
	AG_InitConfigWin(0);
	AG_SetRefreshRate(fps);
	AG_BindGlobalKey(SDLK_ESCAPE, KMOD_NONE, AG_Quit);
	AG_BindGlobalKey(SDLK_F1, KMOD_NONE, AG_ShowSettings);
	AG_BindGlobalKey(SDLK_F8, KMOD_NONE, AG_ViewCapture);
	
	CreateTextbox();

	AG_EventLoop();
	AG_Destroy();
	return (0);
fail:
	AG_Destroy();
	return (1);
}

